.circleci/config.yml

```diff
@@ -1,108 +1,132 @@
-            python --version
-            python -m unittest discover
+      - ocean/pyenv-install-python:
+          python-version: << parameters.python-version >>
+          cache: true
 
-  test-3.8:
-    <<: *full-test-template
-    docker:
-      - image: circleci/python:3.8
+      - ocean/pip-install:
+          requirements: tests/requirements.txt
+          cache: true
 
-  test-3.7:
-    <<: *full-test-template
-    docker:
-      - image: circleci/python:3.7
+      - ocean/pip-install:
+          packages: .
+          cache: false
 
-  test-3.6:
-    <<: *full-test-template
-    docker:
-      - image: circleci/python:3.6
+      - ocean/coverage-run-unittest
 
-  test-osx-3.9: &osx-tests-template
-    macos:
-      xcode: "12.2.0"
-    environment:
-      PYTHON: 3.9.0
-      HOMEBREW_NO_AUTO_UPDATE: 1
+  test-individual-pkgs:
+    docker:
+      - image: circleci/python:3.9
 
     working_directory: ~/repo
 
-    steps: 
+    steps:
       - checkout
 
-      - run: 
-          name: install pyenv
+      - run: &initialize-submodules
+          name: initialize submodules
           command: |
-            brew install pyenv
+            git submodule init
+            git submodule update
 
-      - restore_cache:
-          key: v2-pyenv-{{ .Environment.CIRCLE_JOB }}-xcode-12.2.0
+      - run: &create-virtualenv
+          name: create virtualenv
+          command: |
+            python -m virtualenv env
 
       - run:
-          name: install python
+          name: install penaltymodel dependencies
           command: |
-            pyenv install $PYTHON -s
-
-      - save_cache:
-          paths:
-            - ~/.pyenv
-          key: v2-pyenv-{{ .Environment.CIRCLE_JOB }}-xcode-12.2.0
+            . env/bin/activate
+            cd penaltymodel
+            pip install -r requirements.txt
+            pip install penaltymodel_core/
+            pip install penaltymodel_cache/
+            pip install penaltymodel_maxgap/
+            pip install penaltymodel_mip/
+            pip install penaltymodel_lp/
+            cd ..
 
       - run:
-          name: create virtualenv
+          name: install dependencies
           command: |
-            eval "$(pyenv init -)"
-            pyenv local $PYTHON
-            python -m pip install virtualenv
-            python -m virtualenv env
-
-      - run: *install-dependencies-template
-        
-      - run: *run-tests-template
-
-  test-osx-3.8:
-    <<: *osx-tests-template
-    environment:
-      PYTHON: 3.8.6
-      HOMEBREW_NO_AUTO_UPDATE: 1
+            . env/bin/activate
+            python --version
+            for submodule in */; do 
+              if [[ -f $submodule/tests/requirements.txt ]]; then
+                pip install -r $submodule/tests/requirements.txt
+              fi
+              if [[ -f $submodule/test_requirements.txt ]]; then
+                pip install -r $submodule/test_requirements.txt
+              fi
+              done
```

Comment:
Oh, on second look, you're installing only *test* requirements here, not package/build requirements, so that should be fine. Test requirements are usually relatively unbounded and thin.